src/polymer/iron-flex-layout-classes.ts

```diff
@@ -0,0 +1,262 @@
+  `).trim();
+
+  async checkDocument(parsedDocument: ParsedHtmlDocument, document: Document) {
+    const warnings: FixableWarning[] = [];
+
+    this.convertDeclarations(parsedDocument, document, warnings);
+
+    return warnings;
+  }
+
+  convertDeclarations(
+      parsedDocument: ParsedHtmlDocument, document: Document,
+      warnings: FixableWarning[]) {
+    // Search in the dom-modules.
+    for (const domModule of document.getFeatures({kind: 'dom-module'})) {
+      const template = dom5.query(domModule.astNode, p.hasTagName('template'));
+      if (!template) {
+        continue;
+      }
+      const templateContent = treeAdapters.default.getTemplateContent(template);
+      const missingModules = getMissingStyleModules(templateContent);
+      if (!missingModules) {
+        continue;
+      }
+      // TODO(valdrin): update the warning location to be at the spot where
+      // the class is used.
+      const warning =
+          createWarning(parsedDocument, domModule.astNode, missingModules);
+      const styleNode = getStyleNodeToEdit(templateContent);
+      if (!styleNode) {
+        const indent = getIndentationInside(templateContent);
+        warning.fix = [{
+          replacementText: `
+${indent}<style include="${missingModules}"></style>`,
+          range: sourceRangeForPrependContent(parsedDocument, template)
+        }];
+      } else if (dom5.hasAttribute(styleNode, 'include')) {
+        warning.fix = [{
+          replacementText: ` ${missingModules}`,
+          range: sourceRangeForAppendAttributeValue(
+              parsedDocument, styleNode, 'include')!
+        }];
+      } else {
+        warning.fix = [{
+          replacementText: ` include="${missingModules}"`,
+          range: sourceRangeForAddAttribute(parsedDocument, styleNode)
+        }];
+      }
+      warnings.push(warning);
+    }
+    const body = dom5.query(parsedDocument.ast, p.hasTagName('body'));
+    // Handle files like `<dom-module></dom-module> <body><p>hello</p></body>`
+    // where a "fake" body node would be created by dom-module. Skip these
+    // cases, dear user please write proper HTML ¯\_(ツ)_/¯
+    if (!body || !body.__location) {
+      return;
+    }
+    const missingModules = getMissingStyleModules(body);
+    if (!missingModules) {
+      return;
+    }
+    // TODO(valdrin): update the warning location to be at the spot where
+    // the class is used.
+    const warning = createWarning(parsedDocument, body, missingModules);
+    const indent = getIndentationInside(body);
+    warning.fix = [{
+      replacementText: `
+${indent}<custom-style>
+${indent}  <style is="custom-style" include="${missingModules}"></style>
+${indent}</custom-style>`,
+      range: sourceRangeForPrependContent(parsedDocument, body)
+    }];
+    warnings.push(warning);
+  }
+}
+
+function getMissingStyleModules(node: dom5.Node) {
+  const usedModules = styleModules.filter((m) => !!dom5.query(node, m.selector))
```

Comment:
This does N passes over the document, one for each entry in `styleModules`. I'd recommend instead using dom5.nodeWalkAll and applying the predicate yourself. Maybe building a set of the modules that match.

Reply:
I've changed this to minimize the queries by first checking the used modules, and then avoid querying nodes if the module is already imported. I'm not sure how nodeWalkAll would save us from reducing the queries..

Reply:
In our experience with the analyzer, code shaped like this:

```
for node in ast:
  for thing_to_do in things_to_do:
    thing_to_do(node)
```

is much faster than

```
for thing_to_do in things_to_do:
  for node in ast:
    thing_to_do(node)
```

This may be less true of HTML than javascript though, as the JS AST is less standard and harder to walk